csclient/csclient.go

```diff
@@ -0,0 +1,487 @@
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot open charm archive")
+	}
+	defer r.Close()
+	return c.uploadArchive(id, r, hash, size)
+}
+
+// UploadBundle uploads the given bundle to the charm store with the given id.
+// The id should include the "bundle" series and should not include the
+// revision. The accepted bundle implementations are charm.BundleDir and
+// charm.BundleArchive.
+func (c *Client) UploadBundle(id *charm.Reference, b charm.Bundle) (*charm.Reference, error) {
+	r, hash, size, err := openArchive(b)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot open bundle archive")
+	}
+	defer r.Close()
+	return c.uploadArchive(id, r, hash, size)
+}
+
+// uploadArchive pushes the archive for the charm or bundle represented by
+// the given body, its SHA384 hash and its size. It returns the resulting
+// entity reference. The given id should include the series and should not
+// include the revision.
+func (c *Client) uploadArchive(id *charm.Reference, body io.ReadSeeker, hash string, size int64) (*charm.Reference, error) {
+	// Validate the entity id.
+	if id.Series == "" {
+		return nil, errgo.Newf("no series specified in %q", id)
+	}
+	if id.Revision != -1 {
+		return nil, errgo.Newf("revision specified in %q, but should not be specified", id)
+	}
+
+	// Prepare the request.
+	req, err := http.NewRequest("POST", "", nil)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot make new request")
+	}
+	req.Header.Set("Content-Type", "application/zip")
+	req.ContentLength = size
+
+	// Send the request.
+	resp, err := c.DoWithBody(req, "/"+id.Path()+"/archive?hash="+hash, httpbakery.SeekerBody(body))
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot post archive")
+	}
+	defer resp.Body.Close()
+
+	// Parse the response.
+	var result params.ArchiveUploadResponse
+	if err := parseResponseBody(resp.Body, &result); err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return result.Id, nil
+}
+
+// PutExtraInfo puts extra-info data for the given id.
+// Each entry in the info map causes a value in extra-info with
+// that key to be set to the associated value.
+// Entries not set in the map will be unchanged.
+func (c *Client) PutExtraInfo(id *charm.Reference, info map[string]interface{}) error {
+	req, _ := http.NewRequest("PUT", "", nil)
+	req.Header.Set("Content-Type", "application/json")
+	data, err := json.Marshal(info)
+	if err != nil {
+		return errgo.Notef(err, "cannot marshal extra-info")
+	}
+	body := bytes.NewReader(data)
+	resp, err := c.DoWithBody(req, "/"+id.Path()+"/meta/extra-info", httpbakery.SeekerBody(body))
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+	resp.Body.Close()
+	return nil
+}
+
+// Meta fetches metadata on the charm or bundle with the
+// given id. The result value provides a value
+// to be filled in with the result, which must be
+// a pointer to a struct containing members corresponding
+// to possible metadata include parameters (see http://tinyurl.com/nysdjly).
```

Comment:
The link is to the charm store API v4 spec. All the charmstore endpoints already have those links.
We should update all of them in a branch and point to the spec included in this repository in github.
But that's a separate task.